### PR TITLE
Improve performance of mergeLocaleMessage (#1099)

### PR DESCRIPTION
**Pull request by cslee**

### CODE REVIEW

File: src/index.js [hunk 1]
```diff
@@ -766,7 +766,12 @@ export default class VueI18n {
     if (this._warnHtmlInMessage === 'warn' || this._warnHtmlInMessage === 'error') {
       this._checkLocaleMessage(locale, this._warnHtmlInMessage, message)
     }
-    this._vm.$set(this._vm.messages, locale, merge({}, this._vm.messages[locale] || {}, message))
+    this._vm.$set(this._vm.messages, locale, merge(
+      typeof this._vm.messages[locale] !== 'undefined' && Object.keys(this._vm.messages[locale]).length
+        ? this._vm.messages[locale]
+        : {},
+      message
+    ))
```
Comment (exoego) [inline src/index.js:774]: Hmm, could you tell me how this change will improve performance when `_vm.messages[locale]` is undefined or empty object ?

This change introduce `Object.keys`, which seems to add extra cost.
I also assume `_vm.messages` is likely to have several messages for many cases, so this change will degrade performance for many cases.

Comment (cslee) [reply src/index.js:774]: Hi exoego,

https://github.com/kazupon/vue-i18n/blob/4b34aa91c46e5c8acd83763da1e03bd428c6a8aa/src/index.js#L769

With reference to the current logic, the `merge` call becomes either `merge({}, this._vm.messages[locale], message)` or `merge({}, {}, message)` when `_vm.messages[locale]` is defined or undefined respectively. The `merge` function loops the messages twice in both cases.

With this change, the messages will only be looped once.

```js
const i18n = new VueI18n({
  locale: 'en',
  messages: {
    ja: { hello: 'world' }
  }
})

const start = process.hrtime()
for (let i = 0; i < 1000000; i++) {
  i18n.mergeLocaleMessage('en', { foo: 'bar' })
}
const end = process.hrtime(start)
console.log(end)

const start2 = process.hrtime()
for (let i = 0; i < 1000000; i++) {
  i18n.mergeLocaleMessage('ja', { foo: 'bar' })
}
const end2 = process.hrtime(start2)
console.log(end2)
```

Comment (exoego) [reply src/index.js:774]: Thanks for explanation and quick benchmark.

> The time difference will be greater when there are more messages.

Yes, this is exactly the case I am afraid in the previous comment: `_vm.messages[locale]` is likely to have several messages for many cases, so this change will degrade performance for many cases.

Could you tell more detail of your use case #1099, where `mergeLocaleMessage` is invoked over undefined or empty message ?

Comment (cslee) [reply src/index.js:774]: Thanks for your review.

> The time difference will be greater when there are more messages.

Sorry for my vague statement, I meant the time difference (before and after the change) will be greater. So the improvement of this push request will be greater when there are more messages to merge which should be the usual cases.

If I replace `{ foo: 'bar' }` in above benchmark with an object of 100 keys. The result becomes:

```js
// Before the change
// [ 10, 646949552 ]
// [ 11, 363784417 ]

// After the change
// [ 0, 882620796 ]
// [ 0, 689321051 ]
```

> Could you tell more detail of your use case #1099, where mergeLocaleMessage is invoked over undefined or empty message ?

I just tried to improve the performance without logic change. The use case was introduced in #752 by another person and was released with version 8.15.3.

Comment (kazupon) [reply src/index.js:774]: @cslee
Thank you for your reporting!
I actually tried to check about performance.

The more locale messages, the more noticeable the performance degradation becomes.
This is a problem.
You can solve the performance degradation with PR.
Thanks! :)

**Review by kazupon — APPROVED** [review]

**Pull request merged**